dev/doc/parsing.md

```diff
@@ -0,0 +1,378 @@
+very specific to Coq (not so similar to Camlp5):
+
+* OCaml code - OCaml code enclosed in curly braces, which is copied verbatim to the generated `*.ml` file
+
+* Comments - comments in the `*.mlg` file in the form `(* ... *)`, which are dropped.  Comments
+  in OCaml code are preserved.
+
+* `DECLARE_PLUGIN "*_plugin"` - associates the file with a specific plugin, for example "ltac_plugin"
+
+* `GRAMMAR EXTEND` - adds additional nonterminals and productions to the grammar and declares global
+  nonterminals referenced in the `GRAMMAR EXTEND`:
+
+  ```
+  GRAMMAR EXTEND Gram
+    GLOBAL:
+      bignat bigint ...;
+    <nonterminal definitions>
+  END
+  ```
+
+  TODO: What is `Gram` above?  Worth explaining?
+
+  Global nonterminals are declared in `pcoq.ml`, e.g. `let bignat = Entry.create "Prim.bignat"`.
+  All the `*.mlg` files include `open Pcoq` and often its modules, e.g. `open Pcoq.Prim`.
+
+* `VERNAC COMMAND EXTEND` - adds new command syntax by adding productions to the
+  `command` nonterminal.  For example:
+
+  ```
+  VERNAC COMMAND EXTEND ExtractionLibrary CLASSIFIED AS QUERY
+  | [ "Extraction" "Library" ident(m) ]
+    -> { extraction_library false m }
+  END
+  ```
+
+  TODO: explain "ExtractionLibrary", CLASSIFIED AS, CLASSIFIED BY, "{ tactic_mode }", STATE
+
+  Productions here are represented with alternate syntax, described later.
+
+  It's also possible to add productions to `command` with `GRAMMAR EXTEND` but
+  without the additional fields.
+
+  TODO: Which is the best practice?
+
+* `VERNAC { ... } EXTEND` - TODO
+
+* `TACTIC EXTEND` - adds new tactic syntax by adding productions to `simple_tactic`.
+  For example:
+
+  ```
+  TACTIC EXTEND btauto
+  | [ "btauto" ] -> { Refl_btauto.Btauto.tac }
+  END
+  ```
+
+  TODO: "btauto", DEPRECATED, LEVEL
+
+* `ARGUMENT EXTEND` - defines a new nonterminal
+
+  ```
+  ARGUMENT EXTEND ast_closure_term
+       PRINTED BY { pp_ast_closure_term }
+       INTERPRETED BY { interp_ast_closure_term }
+       GLOBALIZED BY { glob_ast_closure_term }
+       SUBSTITUTED BY { subst_ast_closure_term }
+       RAW_PRINTED BY { pp_ast_closure_term }
+       GLOB_PRINTED BY { pp_ast_closure_term }
+    | [ term_annotation(a) constr(c) ] -> { mk_ast_closure_term a c }
+  END
+  ```
+
+  TODO: what is the purpose of this?
+  TODO: what are the various keywords for?  And TYPED AS?
+
+  Creates a "wit_" symbol
+
+* `VERNAC ARGUMENT EXTEND` - (part of `argument_extend` in the yacc grammar) defines
+  productions for a single nonterminal.  For example:
+
+  ```
+  VERNAC ARGUMENT EXTEND language
+  PRINTED BY { pr_language }
+  | [ "Ocaml" ] -> { let _ = warn_deprecated_ocaml_spelling () in Ocaml }
+  | [ "OCaml" ] -> { Ocaml }
+  | [ "Haskell" ] -> { Haskell }
+  | [ "Scheme" ] -> { Scheme }
+  | [ "JSON" ] -> { JSON }
+  END
+  ```
+
+  TODO: creates wit_ variable--makes it possible to refer to when using alternative prodn syntax?
+
+  TODO: best practice - use this or GRAMMAR EXTEND?
```

Comment:
Don't use GRAMMAR EXTEND, except when you know what you're doing, which is the unique privilege of ssr developers (and even them, sometimes they do strange stuff).

Reply:
There must be more to the story.  I see GRAMMAR EXTEND is used in 17 mlg files.  Is part of this tech debt (e.g. in `g_indfun.mlg`)?  If so, I can say that.  Also I can say that new tactics and commands should not be added with GRAMMAR EXTEND.

Reply:
The funind instance looks like an oversight. This should be turned into a `VERNAC ARGUMENT EXTEND` since there is not much more to it.

The typical use of `GRAMMAR EXTEND` is for low-level complex syntactic constructs, typically when implementing a full-blown language grammar (e.g. gallina, LtacX, ssr). Not for the casual user, thus.

Reply:
OK.  The commands in `g_proofs.mlg`, `g_vernac.mlg` and `g_ltac.mlg`,  inside `GRAMMAR EXTEND` are OK because they're big extensions?

Reply:
In a nutshell, yes.